File: misc/dumb.c
#include <assert.h>
#include <errno.h>
#include <getopt.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/ioctl.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <unistd.h>

#define MAXSIG 31
char signal_temporary_ignores[MAXSIG + 1] = {[0 ... MAXSIG] = 0};
pid_t child_pid = -1;

void handle_signal(int signum) {
  fprintf(stderr, "signal %i\n", signum);
  if (signal_temporary_ignores[signum] == 1) {
    signal_temporary_ignores[signum] = 0;
  } else if (signum == SIGCHLD) {
    int status, exit_status;
    pid_t killed_pid;
    while ((killed_pid = waitpid(-1, &status, WNOHANG)) > 0) {
      if (WIFEXITED(status)) {
        exit_status = WEXITSTATUS(status);
      } else {
        assert(WIFSIGNALED(status));
        exit_status = 128 + WTERMSIG(status);
      }
      if (killed_pid == child_pid) {
        kill(-child_pid, SIGTERM);
        exit(exit_status);
      }
    }
  } else {
    if (signum != 0) kill(-child_pid, signum);
    if (signum == SIGTSTP || signum == SIGTTOU || signum == SIGTTIN) {
      kill(getpid(), SIGSTOP);
    }
  }
}

void dummy(int signum) {}

int dumb (int argc, char * argv[]) {
  sigset_t all_signals;
  sigfillset(&all_signals);
  sigprocmask(SIG_BLOCK, &all_signals, NULL);
  int i = 0;
  for (i = 1; i <= MAXSIG; i++) {
    signal(i, dummy);
  }
  if (ioctl(STDIN_FILENO, TIOCNOTTY) != -1) {
    if (getsid(0) == getpid()) {
      signal_temporary_ignores[SIGHUP] = 1;
      signal_temporary_ignores[SIGCONT] = 1;
    }
  }
  child_pid = fork();
  if (child_pid < 0) {
    return 1;
  } else if (child_pid == 0) {
    /* child */
    sigprocmask(SIG_UNBLOCK, &all_signals, NULL);
    if (setsid() == -1) {
      exit(1);
    }
    ioctl(STDIN_FILENO, TIOCSCTTY, 0);
    execvp("/usr/local/bin/just", &argv[0]);
    return 2;
  } else {
    /* parent */
    if (chdir("/") == -1) {}
    for (;;) {
      int signum;
      sigwait(&all_signals, &signum);
      handle_signal(signum);
    }
  }
}

int main(int argc, char *argv[]) {
  return dumb(argc, argv);
}

File: misc/dumb.js
const { signal } = just.library('signal')
const { sys } = just.library('sys')

function onSignal (signum) {
  if (ignore[signum]) {
    ignore[signum] = 0
    return
  }
  let exitStatus = 0
  if (signum === signal.SIGCHLD) {
    let [status, kpid] = sys.waitpid(new Uint32Array(2))
    while (kpid > 0) {
      if ((status & 0x7f) === 0) { // WIFEXITED
        exitStatus = ((status & 0xff00) >> 8) // WEXITSTATUS
      } else {
        // assert(WIFSIGNALED(status));
        exitStatus = 128 + (status & 0x7f) // WTERMSIG
      }
      if (kpid === child) {
        sys.kill(-child, signal.SIGTERM)
        sys.exit(exitStatus)
      }
      [status, kpid] = sys.waitpid(new Uint32Array(2))
    }
  }
  if (signum !== 0) sys.kill(-child, signum)
  if (signum === signal.SIGTSTP ||
    signum === signal.SIGTTOU ||
    signum === signal.SIGTTIN) {
    sys.kill(sys.pid(), signal.SIGSTOP)
  }
}

function parentMain () {
  signal.reset()
  just.chdir('/')
  for (let i = 1; i <= MAXSIG; i++) {
    signal.sigaction(i, onSignal)
  }
  while (1) just.sleep(1)
}

function childMain () {
  signal.sigprocmask(sigmask, signal.SIG_UNBLOCK, 1)
  if (sys.setsid() === -1) {
    sys.exit(1)
  }
  sys.ioctl(sys.STDIN_FILENO, TIOCSCTTY)
  sys.exec(just.args[1], just.args.slice(2))
  sys.exit(2)
}

const ignore = new Array(32)
const MAXSIG = 31
const TIOCNOTTY = 0x5422
const TIOCSCTTY = 0x540E
const sigmask = new ArrayBuffer(128)
signal.SIGTTIN = 21
signal.SIGTTOU = 22
ignore.fill(0)
signal.sigfillset(sigmask)
signal.sigprocmask(sigmask, signal.SIG_BLOCK, 1)
if (sys.ioctl(sys.STDIN_FILENO, TIOCNOTTY) !== -1) {
  if (sys.getsid(0) === sys.pid()) {
    ignore[signal.SIGHUP] = 1
    ignore[signal.SIGCONT] = 1
  }
}
const child = sys.fork()
if (child < 0) sys.exit(1)
child === 0 ? childMain() : parentMain()
